rb/lib/selenium/webdriver/remote/response.rb

```diff
@@ -59,34 +56,12 @@ def assert_ok
           raise Error::ServerError, self
         end
 
-        def add_backtrace(ex, server_trace)
-          return unless server_trace
-
-          backtrace = case server_trace
-                      when Array
-                        backtrace_from_remote(server_trace)
-                      when String
-                        server_trace.split("\n")
-                      end
-
-          ex.set_backtrace(backtrace + ex.backtrace)
-        end
-
-        def backtrace_from_remote(server_trace)
```

Comment:
I don't think you need to drop this code. Your implementation works fine for a local driver scenario (e.g. chromedriver) but when used with Selenium Server/Grid, you need this extra logic to process and clean the stack trace coming from the server.

Can you try experimenting with `Selenium::WebDriver.for(:remote, options: Selenium::WebDriver::Options.chrome)` and share how the error cause looks like?

Reply:
I have been experimenting and when doing:

```
    context 'with a remote driver', only: [{driver: :remote}] do
     it 'has cause' do
          driver_test = Selenium::WebDriver.for(:remote, options: Selenium::WebDriver::Options.chrome)
          driver_test.find_element(id: 'nonexistent')
        rescue WebDriver::Error::NoSuchElementError => e
          pp e.cause
          expect(e.cause).to be_a(WebDriver::Error::WebDriverError)
          driver.quit
        end
      end
```

Without the backtrace_from_remote and add_backtrace method cause looks like:

```
#<Selenium::WebDriver::Error::WebDriverError: 0   chromedriver                        0x00000001052bea80 chromedriver + 4385408
1   chromedriver                        0x00000001052b738c chromedriver + 4354956
2   chromedriver                        0x0000000104ed4b0c chromedriver + 281356
3   chromedriver                        0x0000000104f172f8 chromedriver + 553720
4   chromedriver                        0x0000000104f4fd24 chromedriver + 785700
5   chromedriver                        0x0000000104f0beec chromedriver + 507628
6   chromedriver                        0x0000000104f0c8c4 chromedriver + 510148
7   chromedriver                        0x000000010528643c chromedriver + 4154428
8   chromedriver                        0x000000010528aea0 chromedriver + 4173472
9   chromedriver                        0x000000010526bff8 chromedriver + 4046840
10  chromedriver                        0x000000010528b78c chromedriver + 4175756
11  chromedriver                        0x000000010525efb8 chromedriver + 3993528
12  chromedriver                        0x00000001052a921c chromedriver + 4297244
13  chromedriver                        0x00000001052a9398 chromedriver + 4297624
14  chromedriver                        0x00000001052b6f84 chromedriver + 4353924
15  libsystem_pthread.dylib             0x00000001940cef94 _pthread_start + 136
16  libsystem_pthread.dylib             0x00000001940c9d34 thread_start + 8
>
```

This is how it looks after adding back set_backtrace and backtrace_from_remote:

```
#<Selenium::WebDriver::Error::WebDriverError: 0   chromedriver                        0x0000000102c02a80 chromedriver + 4385408
1   chromedriver                        0x0000000102bfb38c chromedriver + 4354956
2   chromedriver                        0x0000000102818b0c chromedriver + 281356
3   chromedriver                        0x000000010285b2f8 chromedriver + 553720
4   chromedriver                        0x0000000102893d24 chromedriver + 785700
5   chromedriver                        0x000000010284feec chromedriver + 507628
6   chromedriver                        0x00000001028508c4 chromedriver + 510148
7   chromedriver                        0x0000000102bca43c chromedriver + 4154428
8   chromedriver                        0x0000000102bceea0 chromedriver + 4173472
9   chromedriver                        0x0000000102bafff8 chromedriver + 4046840
10  chromedriver                        0x0000000102bcf78c chromedriver + 4175756
11  chromedriver                        0x0000000102ba2fb8 chromedriver + 3993528
12  chromedriver                        0x0000000102bed21c chromedriver + 4297244
13  chromedriver                        0x0000000102bed398 chromedriver + 4297624
14  chromedriver                        0x0000000102bfaf84 chromedriver + 4353924
15  libsystem_pthread.dylib             0x00000001940cef94 _pthread_start + 136
16  libsystem_pthread.dylib             0x00000001940c9d34 thread_start + 8
```

Reply:
So the results is pretty similar in both cases, and since I have to pass the backtrace as a string for the second exception with the add backtrace method it always going to just use the case where server_trace is a string unless I force it to be an array, so I do not see too much of a difference in the case of having a remote driver

But maybe I'm missing something @p0deje

Reply:
@aguspe You likely need to use a command that invokes server-specific command and check that stacktrace is preserved. For example, If you run the following in master, you can see the server stacktrace:

```ruby
driver = Selenium::WebDriver.for(:remote, options: Selenium::WebDriver::Options.chrome)
driver.send(:bridge).downloadable_files
# org.openqa.selenium.WebDriverException: Please enable management of downloads via the command line arg [--enable-managed-downloads] and restart the node: Please enable management of downloads via the command line arg [--enable-managed-downloads] and restart the node (Selenium::WebDriver::Error::UnknownError)
# Build info: version: '4.23.0-SNAPSHOT', revision: 'Unknown'
# System info: os.name: 'Mac OS X', os.arch: 'aarch64', os.version: '14.6', java.version: '21.0.2'
# Driver info: driver.version: unknown
# 	from Build info: version: '4.23.0-SNAPSHOT', revision: 'Unknown'
# 	from System info: os.name: 'Mac OS X', os.arch: 'aarch64', os.version: '14.6', java.version: '21.0.2'
# 	from Driver info: driver.version: unknown
# 	from 	at org.openqa.selenium.grid.node.local.LocalNode.downloadFile(LocalNode.java:652)
# 	from 	at org.openqa.selenium.grid.node.DownloadFile.execute(DownloadFile.java:39)
# 	from 	at org.openqa.selenium.remote.http.Route$TemplatizedRoute.handle(Route.java:192)
# 	from 	at org.openqa.selenium.remote.http.Route.execute(Route.java:69)
# 	from 	at org.openqa.selenium.remote.tracing.SpanWrappedHttpHandler.execute(SpanWrappedHttpHandler.java:87)
# 	from 	at org.openqa.selenium.remote.http.Filter$1.execute(Filter.java:63)
# 	from 	at org.openqa.selenium.remote.http.Route$CombinedRoute.handle(Route.java:360)
# 	from 	at org.openqa.selenium.remote.http.Route.execute(Route.java:69)
# 	from 	at org.openqa.selenium.grid.node.Node.execute(Node.java:270)
# 	from 	at org.openqa.selenium.grid.web.CombinedHandler.execute(CombinedHandler.java:55)
# 	from 	at org.openqa.selenium.grid.web.RoutableHttpClientFactory$1.execute(RoutableHttpClientFactory.java:71)
# 	from 	at org.openqa.selenium.grid.web.ReverseProxyHandler.execute(ReverseProxyHandler.java:90)
# 	from 	at org.openqa.selenium.grid.router.HandleSession.execute(HandleSession.java:181)
# ...
```

This stacktrace should now go in `cause` in your PR.

Reply:
Thank you for the guidance here @p0deje, I tried 3 different things using this code snippet:

```Ruby
      it 'has cause' do
        driver = Selenium::WebDriver.for(:remote, options: Selenium::WebDriver::Options.chrome)
        driver.send(:bridge).downloadable_files
      rescue WebDriver::Error::UnknownError => e
        pp e.cause
        expect(e.backtrace).not_to be_empty
        driver.quit
      end
```

On this branch without the add_backtrace and backtrace_from_remote methods, cause resulted in:

```Ruby
#<Selenium::WebDriver::Error::WebDriverError: org.openqa.selenium.WebDriverException: Please enable management of downloads via the command line arg [--enable-managed-downloads] and restart the node
Build info: version: '4.23.0-SNAPSHOT', revision: '5494108'
System info: os.name: 'Mac OS X', os.arch: 'aarch64', os.version: '14.5', java.version: '21.0.2'
Driver info: driver.version: unknown
	at org.openqa.selenium.grid.node.local.LocalNode.downloadFile(LocalNode.java:652)
	at org.openqa.selenium.grid.node.DownloadFile.execute(DownloadFile.java:39)
	at org.openqa.selenium.remote.http.Route$TemplatizedRoute.handle(Route.java:192)
	at org.openqa.selenium.remote.http.Route.execute(Route.java:69)
	at org.openqa.selenium.remote.tracing.SpanWrappedHttpHandler.execute(SpanWrappedHttpHandler.java:87)
	at org.openqa.selenium.remote.http.Filter$1.execute(Filter.java:63)
	at org.openqa.selenium.remote.http.Route$CombinedRoute.handle(Route.java:360)
	at org.openqa.selenium.remote.http.Route.execute(Route.java:69)
	at org.openqa.selenium.grid.node.Node.execute(Node.java:270)
	at org.openqa.selenium.grid.web.CombinedHandler.execute(CombinedHandler.java:55)
	at org.openqa.selenium.grid.web.RoutableHttpClientFactory$1.execute(RoutableHttpClientFactory.java:71)
	at org.openqa.selenium.grid.web.ReverseProxyHandler.execute(ReverseProxyHandler.java:90)
	at org.openqa.selenium.grid.router.HandleSession.execute(HandleSession.java:181)
	at org.openqa.selenium.remote.http.Route$PredicatedRoute.handle(Route.java:397)
	at org.openqa.selenium.remote.http.Route.execute(Route.java:69)
	at org.openqa.selenium.remote.http.Route$CombinedRoute.handle(Route.java:360)
	at org.openqa.selenium.remote.http.Route.execute(Route.java:69)
	at org.openqa.selenium.grid.router.Router.execute(Router.java:87)
	at org.openqa.selenium.grid.web.CheckOriginHeader.lambda$apply$0(CheckOriginHeader.java:69)
	at org.openqa.selenium.grid.web.CheckContentTypeHeader.lambda$apply$0(CheckContentTypeHeader.java:75)
	at org.openqa.selenium.grid.web.EnsureSpecCompliantResponseHeaders.lambda$apply$0(EnsureSpecCompliantResponseHeaders.java:34)
	at org.openqa.selenium.remote.http.Filter$1.execute(Filter.java:63)
	at org.openqa.selenium.remote.http.Route$CombinedRoute.handle(Route.java:360)
	at org.openqa.selenium.remote.http.Route.execute(Route.java:69)
	at org.openqa.selenium.remote.http.Route$NestedRoute.handle(Route.java:270)
	at org.openqa.selenium.remote.http.Route.execute(Route.java:69)
	at org.openqa.selenium.remote.http.Route$CombinedRoute.handle(Route.java:360)
	at org.openqa.selenium.remote.http.Route.execute(Route.java:69)
	at org.openqa.selenium.remote.AddWebDriverSpecHeaders.lambda$apply$0(AddWebDriverSpecHeaders.java:35)
	at org.openqa.selenium.remote.ErrorFilter.lambda$apply$0(ErrorFilter.java:44)
	at org.openqa.selenium.remote.http.Filter$1.execute(Filter.java:63)
	at org.openqa.selenium.remote.ErrorFilter.lambda$apply$0(ErrorFilter.java:44)
	at org.openqa.selenium.remote.http.Filter$1.execute(Filter.java:63)
	at org.openqa.selenium.netty.server.SeleniumHandler.lambda$channelRead0$0(SeleniumHandler.java:44)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:572)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base/java.lang.Thread.run(Thread.java:1583)
>
```

If we add the methods cause looks like this:

```Ruby
#<Selenium::WebDriver::Error::WebDriverError: ["org.openqa.selenium.WebDriverException: Please enable management of downloads via the command line arg [--enable-managed-downloads] and restart the node", "Build info: version: '4.23.0-SNAPSHOT', revision: '5494108'", "System info: os.name: 'Mac OS X', os.arch: 'aarch64', os.version: '14.5', java.version: '21.0.2'", "Driver info: driver.version: unknown", "\tat org.openqa.selenium.grid.node.local.LocalNode.downloadFile(LocalNode.java:652)", "\tat org.openqa.selenium.grid.node.DownloadFile.execute(DownloadFile.java:39)", "\tat org.openqa.selenium.remote.http.Route$TemplatizedRoute.handle(Route.java:192)", "\tat org.openqa.selenium.remote.http.Route.execute(Route.java:69)", "\tat org.openqa.selenium.remote.tracing.SpanWrappedHttpHandler.execute(SpanWrappedHttpHandler.java:87)", "\tat org.openqa.selenium.remote.http.Filter$1.execute(Filter.java:63)", "\tat org.openqa.selenium.remote.http.Route$CombinedRoute.handle(Route.java:360)", "\tat org.openqa.selenium.remote.http.Route.execute(Route.java:69)", "\tat org.openqa.selenium.grid.node.Node.execute(Node.java:270)", "\tat org.openqa.selenium.grid.web.CombinedHandler.execute(CombinedHandler.java:55)", "\tat org.openqa.selenium.grid.web.RoutableHttpClientFactory$1.execute(RoutableHttpClientFactory.java:71)", "\tat org.openqa.selenium.grid.web.ReverseProxyHandler.execute(ReverseProxyHandler.java:90)", "\tat org.openqa.selenium.grid.router.HandleSession.execute(HandleSession.java:181)", "\tat org.openqa.selenium.remote.http.Route$PredicatedRoute.handle(Route.java:397)", "\tat org.openqa.selenium.remote.http.Route.execute(Route.java:69)", "\tat org.openqa.selenium.remote.http.Route$CombinedRoute.handle(Route.java:360)", "\tat org.openqa.selenium.remote.http.Route.execute(Route.java:69)", "\tat org.openqa.selenium.grid.router.Router.execute(Router.java:87)", "\tat org.openqa.selenium.grid.web.CheckOriginHeader.lambda$apply$0(CheckOriginHeader.java:69)", "\tat org.openqa.selenium.grid.web.CheckContentTypeHeader.lambda$apply$0(CheckContentTypeHeader.java:75)", "\tat org.openqa.selenium.grid.web.EnsureSpecCompliantResponseHeaders.lambda$apply$0(EnsureSpecCompliantResponseHeaders.java:34)", "\tat org.openqa.selenium.remote.http.Filter$1.execute(Filter.java:63)", "\tat org.openqa.selenium.remote.http.Route$CombinedRoute.handle(Route.java:360)", "\tat org.openqa.selenium.remote.http.Route.execute(Route.java:69)", "\tat org.openqa.selenium.remote.http.Route$NestedRoute.handle(Route.java:270)", "\tat org.openqa.selenium.remote.http.Route.execute(Route.java:69)", "\tat org.openqa.selenium.remote.http.Route$CombinedRoute.handle(Route.java:360)", "\tat org.openqa.selenium.remote.http.Route.execute(Route.java:69)", "\tat org.openqa.selenium.remote.AddWebDriverSpecHeaders.lambda$apply$0(AddWebDriverSpecHeaders.java:35)", "\tat org.openqa.selenium.remote.ErrorFilter.lambda$apply$0(ErrorFilter.java:44)", "\tat org.openqa.selenium.remote.http.Filter$1.execute(Filter.java:63)", "\tat org.openqa.selenium.remote.ErrorFilter.lambda$apply$0(ErrorFilter.java:44)", "\tat org.openqa.selenium.remote.http.Filter$1.execute(Filter.java:63)", "\tat org.openqa.selenium.netty.server.SeleniumHandler.lambda$channelRead0$0(SeleniumHandler.java:44)", "\tat java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:572)", "\tat java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)", "\tat java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)", "\tat java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)", "\tat java.base/java.lang.Thread.run(Thread.java:1583)"]>
```

On the master branch cause is nil, but if we replace e.cause for e.backtrace the result is the following:

```Ruby
["org.openqa.selenium.WebDriverException: Please enable management of downloads via the command line arg [--enable-managed-downloads] and restart the node",
 "Build info: version: '4.23.0-SNAPSHOT', revision: '5494108'",
 "System info: os.name: 'Mac OS X', os.arch: 'aarch64', os.version: '14.5', java.version: '21.0.2'",
 "Driver info: driver.version: unknown",
 "\tat org.openqa.selenium.grid.node.local.LocalNode.downloadFile(LocalNode.java:652)",
 "\tat org.openqa.selenium.grid.node.DownloadFile.execute(DownloadFile.java:39)",
 "\tat org.openqa.selenium.remote.http.Route$TemplatizedRoute.handle(Route.java:192)",
 "\tat org.openqa.selenium.remote.http.Route.execute(Route.java:69)",
 "\tat org.openqa.selenium.remote.tracing.SpanWrappedHttpHandler.execute(SpanWrappedHttpHandler.java:87)",
 "\tat org.openqa.selenium.remote.http.Filter$1.execute(Filter.java:63)",
 "\tat org.openqa.selenium.remote.http.Route$CombinedRoute.handle(Route.java:360)",
 "\tat org.openqa.selenium.remote.http.Route.execute(Route.java:69)",
 "\tat org.openqa.selenium.grid.node.Node.execute(Node.java:270)",
 "\tat org.openqa.selenium.grid.web.CombinedHandler.execute(CombinedHandler.java:55)",
 "\tat org.openqa.selenium.grid.web.RoutableHttpClientFactory$1.execute(RoutableHttpClientFactory.java:71)",
 "\tat org.openqa.selenium.grid.web.ReverseProxyHandler.execute(ReverseProxyHandler.java:90)",
 "\tat org.openqa.selenium.grid.router.HandleSession.execute(HandleSession.java:181)",
 "\tat org.openqa.selenium.remote.http.Route$PredicatedRoute.handle(Route.java:397)",
 "\tat org.openqa.selenium.remote.http.Route.execute(Route.java:69)",
 "\tat org.openqa.selenium.remote.http.Route$CombinedRoute.handle(Route.java:360)",
 "\tat org.openqa.selenium.remote.http.Route.execute(Route.java:69)",
 "\tat org.openqa.selenium.grid.router.Router.execute(Router.java:87)",
 "\tat org.openqa.selenium.grid.web.CheckOriginHeader.lambda$apply$0(CheckOriginHeader.java:69)",
 "\tat org.openqa.selenium.grid.web.CheckContentTypeHeader.lambda$apply$0(CheckContentTypeHeader.java:75)",
 "\tat org.openqa.selenium.grid.web.EnsureSpecCompliantResponseHeaders.lambda$apply$0(EnsureSpecCompliantResponseHeaders.java:34)",
 "\tat org.openqa.selenium.remote.http.Filter$1.execute(Filter.java:63)",
 "\tat org.openqa.selenium.remote.http.Route$CombinedRoute.handle(Route.java:360)",
 "\tat org.openqa.selenium.remote.http.Route.execute(Route.java:69)",
 "\tat org.openqa.selenium.remote.http.Route$NestedRoute.handle(Route.java:270)",
 "\tat org.openqa.selenium.remote.http.Route.execute(Route.java:69)",
 "\tat org.openqa.selenium.remote.http.Route$CombinedRoute.handle(Route.java:360)",
 "\tat org.openqa.selenium.remote.http.Route.execute(Route.java:69)",
 "\tat org.openqa.selenium.remote.AddWebDriverSpecHeaders.lambda$apply$0(AddWebDriverSpecHeaders.java:35)",
 "\tat org.openqa.selenium.remote.ErrorFilter.lambda$apply$0(ErrorFilter.java:44)",
 "\tat org.openqa.selenium.remote.http.Filter$1.execute(Filter.java:63)",
 "\tat org.openqa.selenium.remote.ErrorFilter.lambda$apply$0(ErrorFilter.java:44)",
 "\tat org.openqa.selenium.remote.http.Filter$1.execute(Filter.java:63)",
 "\tat org.openqa.selenium.netty.server.SeleniumHandler.lambda$channelRead0$0(SeleniumHandler.java:44)",
 "\tat java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:572)",
 "\tat java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)",
 "\tat java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)",
 "\tat java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)",
 "\tat java.base/java.lang.Thread.run(Thread.java:1583)",
 "/Users/apeque01/Desktop/main_folder/Projects/open_source/selenium/rb/lib/selenium/webdriver/remote/response.rb:55:in `assert_ok'",
 "/Users/apeque01/Desktop/main_folder/Projects/open_source/selenium/rb/lib/selenium/webdriver/remote/response.rb:34:in `initialize'",
 "/Users/apeque01/Desktop/main_folder/Projects/open_source/selenium/rb/lib/selenium/webdriver/remote/http/common.rb:101:in `new'",
 "/Users/apeque01/Desktop/main_folder/Projects/open_source/selenium/rb/lib/selenium/webdriver/remote/http/common.rb:101:in `create_response'",
 "/Users/apeque01/Desktop/main_folder/Projects/open_source/selenium/rb/lib/selenium/webdriver/remote/http/default.rb:103:in `request'",
 "/Users/apeque01/Desktop/main_folder/Projects/open_source/selenium/rb/lib/selenium/webdriver/remote/http/common.rb:67:in `call'",
 "/Users/apeque01/Desktop/main_folder/Projects/open_source/selenium/rb/lib/selenium/webdriver/remote/bridge.rb:675:in `execute'",
 "/Users/apeque01/Desktop/main_folder/Projects/open_source/selenium/rb/lib/selenium/webdriver/remote/features.rb:62:in `downloadable_files'",
 "/Users/apeque01/Desktop/main_folder/Projects/open_source/selenium/rb/spec/integration/selenium/webdriver/error_spec.rb:35:in `block (2 levels) in <module:WebDriver>'",
 "/Users/apeque01/.rvm/gems/ruby-3.0.0/gems/rspec-core-3.13.0/lib/rspec/core/example.rb:263:in `instance_exec'",
 "/Users/apeque01/.rvm/gems/ruby-3.0.0/gems/rspec-core-3.13.0/lib/rspec/core/example.rb:263:in `block in run'",
 "/Users/apeque01/.rvm/gems/ruby-3.0.0/gems/rspec-core-3.13.0/lib/rspec/core/example.rb:511:in `block in with_around_and_singleton_context_hooks'",
 "/Users/apeque01/.rvm/gems/ruby-3.0.0/gems/rspec-core-3.13.0/lib/rspec/core/example.rb:468:in `block in with_around_example_hooks'",
 "/Users/apeque01/.rvm/gems/ruby-3.0.0/gems/rspec-core-3.13.0/lib/rspec/core/hooks.rb:486:in `block in run'",
 "/Users/apeque01/.rvm/gems/ruby-3.0.0/gems/rspec-core-3.13.0/lib/rspec/core/hooks.rb:626:in `block in run_around_example_hooks_for'",
 "/Users/apeque01/.rvm/gems/ruby-3.0.0/gems/rspec-core-3.13.0/lib/rspec/core/example.rb:352:in `call'",
 "/Users/apeque01/.rvm/gems/ruby-3.0.0/gems/rspec-expectations-3.13.1/lib/rspec/expectations/failure_aggregator.rb:25:in `block in aggregate'",
 "/Users/apeque01/.rvm/gems/ruby-3.0.0/gems/rspec-support-3.13.1/lib/rspec/support.rb:126:in `with_failure_notifier'",
 "/Users/apeque01/.rvm/gems/ruby-3.0.0/gems/rspec-expectations-3.13.1/lib/rspec/expectations/failure_aggregator.rb:23:in `aggregate'",
 "/Users/apeque01/.rvm/gems/ruby-3.0.0/gems/rspec-expectations-3.13.1/lib/rspec/matchers.rb:306:in `aggregate_failures'",
 "/Users/apeque01/.rvm/gems/ruby-3.0.0/gems/rspec-core-3.13.0/lib/rspec/core/configuration.rb:2281:in `block in define_built_in_hooks'",
 "/Users/apeque01/.rvm/gems/ruby-3.0.0/gems/rspec-core-3.13.0/lib/rspec/core/example.rb:457:in `instance_exec'",
 "/Users/apeque01/.rvm/gems/ruby-3.0.0/gems/rspec-core-3.13.0/lib/rspec/core/example.rb:457:in `instance_exec'",
 "/Users/apeque01/.rvm/gems/ruby-3.0.0/gems/rspec-core-3.13.0/lib/rspec/core/hooks.rb:390:in `execute_with'",
 "/Users/apeque01/.rvm/gems/ruby-3.0.0/gems/rspec-core-3.13.0/lib/rspec/core/hooks.rb:628:in `block (2 levels) in run_around_example_hooks_for'",
 "/Users/apeque01/.rvm/gems/ruby-3.0.0/gems/rspec-core-3.13.0/lib/rspec/core/example.rb:352:in `call'",
 "/Users/apeque01/.rvm/gems/ruby-3.0.0/gems/rspec-core-3.13.0/lib/rspec/core/hooks.rb:629:in `run_around_example_hooks_for'",
 "/Users/apeque01/.rvm/gems/ruby-3.0.0/gems/rspec-core-3.13.0/lib/rspec/core/hooks.rb:486:in `run'",
 "/Users/apeque01/.rvm/gems/ruby-3.0.0/gems/rspec-core-3.13.0/lib/rspec/core/example.rb:468:in `with_around_example_hooks'",
 "/Users/apeque01/.rvm/gems/ruby-3.0.0/gems/rspec-core-3.13.0/lib/rspec/core/example.rb:511:in `with_around_and_singleton_context_hooks'",
 "/Users/apeque01/.rvm/gems/ruby-3.0.0/gems/rspec-core-3.13.0/lib/rspec/core/example.rb:259:in `run'",
 "/Users/apeque01/.rvm/gems/ruby-3.0.0/gems/rspec-core-3.13.0/lib/rspec/core/example_group.rb:646:in `block in run_examples'",
 "/Users/apeque01/.rvm/gems/ruby-3.0.0/gems/rspec-core-3.13.0/lib/rspec/core/example_group.rb:642:in `map'",
 "/Users/apeque01/.rvm/gems/ruby-3.0.0/gems/rspec-core-3.13.0/lib/rspec/core/example_group.rb:642:in `run_examples'",
 "/Users/apeque01/.rvm/gems/ruby-3.0.0/gems/rspec-core-3.13.0/lib/rspec/core/example_group.rb:607:in `run'",
 "/Users/apeque01/.rvm/gems/ruby-3.0.0/gems/rspec-core-3.13.0/lib/rspec/core/runner.rb:121:in `block (3 levels) in run_specs'",
 "/Users/apeque01/.rvm/gems/ruby-3.0.0/gems/rspec-core-3.13.0/lib/rspec/core/runner.rb:121:in `map'",
 "/Users/apeque01/.rvm/gems/ruby-3.0.0/gems/rspec-core-3.13.0/lib/rspec/core/runner.rb:121:in `block (2 levels) in run_specs'",
 "/Users/apeque01/.rvm/gems/ruby-3.0.0/gems/rspec-core-3.13.0/lib/rspec/core/configuration.rb:2091:in `with_suite_hooks'",
 "/Users/apeque01/.rvm/gems/ruby-3.0.0/gems/rspec-core-3.13.0/lib/rspec/core/runner.rb:116:in `block in run_specs'",
 "/Users/apeque01/.rvm/gems/ruby-3.0.0/gems/rspec-core-3.13.0/lib/rspec/core/reporter.rb:74:in `report'",
 "/Users/apeque01/.rvm/gems/ruby-3.0.0/gems/rspec-core-3.13.0/lib/rspec/core/runner.rb:115:in `run_specs'",
 "/Users/apeque01/.rvm/gems/ruby-3.0.0/gems/rspec-core-3.13.0/lib/rspec/core/runner.rb:89:in `run'",
 "/Users/apeque01/.rvm/gems/ruby-3.0.0/gems/rspec-core-3.13.0/lib/rspec/core/runner.rb:71:in `run'",
 "/Users/apeque01/.rvm/gems/ruby-3.0.0/gems/rspec-core-3.13.0/lib/rspec/core/runner.rb:45:in `invoke'",
 "/Users/apeque01/.rvm/gems/ruby-3.0.0/gems/rspec-core-3.13.0/exe/rspec:4:in `<top (required)>'",
 "/Users/apeque01/.rvm/gems/ruby-3.0.0/bin/rspec:25:in `load'",
 "/Users/apeque01/.rvm/gems/ruby-3.0.0/bin/rspec:25:in `<main>'"]
```

Reply:
So personally I think we should remove the methods, since the error we pass is always going to be a string, causing the backtrace_from_remote not to be called

the backtrace looks like this branch without the add_backtrace and backtrace_from_remote methods:

```Ruby
["/Users/apeque01/Desktop/main_folder/Projects/open_source/selenium/rb/lib/selenium/webdriver/remote/response.rb:63:in `rescue in add_cause'",
 "/Users/apeque01/Desktop/main_folder/Projects/open_source/selenium/rb/lib/selenium/webdriver/remote/response.rb:60:in `add_cause'",
 "/Users/apeque01/Desktop/main_folder/Projects/open_source/selenium/rb/lib/selenium/webdriver/remote/response.rb:42:in `error'",
 "/Users/apeque01/Desktop/main_folder/Projects/open_source/selenium/rb/lib/selenium/webdriver/remote/response.rb:53:in `assert_ok'",
 "/Users/apeque01/Desktop/main_folder/Projects/open_source/selenium/rb/lib/selenium/webdriver/remote/response.rb:34:in `initialize'",
 "/Users/apeque01/Desktop/main_folder/Projects/open_source/selenium/rb/lib/selenium/webdriver/remote/http/common.rb:101:in `new'",
 "/Users/apeque01/Desktop/main_folder/Projects/open_source/selenium/rb/lib/selenium/webdriver/remote/http/common.rb:101:in `create_response'",
 "/Users/apeque01/Desktop/main_folder/Projects/open_source/selenium/rb/lib/selenium/webdriver/remote/http/default.rb:103:in `request'",
 "/Users/apeque01/Desktop/main_folder/Projects/open_source/selenium/rb/lib/selenium/webdriver/remote/http/common.rb:67:in `call'",
 "/Users/apeque01/Desktop/main_folder/Projects/open_source/selenium/rb/lib/selenium/webdriver/remote/bridge.rb:675:in `execute'",
 "/Users/apeque01/Desktop/main_folder/Projects/open_source/selenium/rb/lib/selenium/webdriver/remote/features.rb:62:in `downloadable_files'",
 "/Users/apeque01/Desktop/main_folder/Projects/open_source/selenium/rb/spec/integration/selenium/webdriver/error_spec.rb:54:in `block (3 levels) in <module:WebDriver>'",
 "/Users/apeque01/.rvm/gems/ruby-3.0.0/gems/rspec-core-3.13.0/lib/rspec/core/example.rb:263:in `instance_exec'",
 "/Users/apeque01/.rvm/gems/ruby-3.0.0/gems/rspec-core-3.13.0/lib/rspec/core/example.rb:263:in `block in run'",
 "/Users/apeque01/.rvm/gems/ruby-3.0.0/gems/rspec-core-3.13.0/lib/rspec/core/example.rb:511:in `block in with_around_and_singleton_context_hooks'",
 "/Users/apeque01/.rvm/gems/ruby-3.0.0/gems/rspec-core-3.13.0/lib/rspec/core/example.rb:468:in `block in with_around_example_hooks'",
 "/Users/apeque01/.rvm/gems/ruby-3.0.0/gems/rspec-core-3.13.0/lib/rspec/core/hooks.rb:486:in `block in run'",
 "/Users/apeque01/.rvm/gems/ruby-3.0.0/gems/rspec-core-3.13.0/lib/rspec/core/hooks.rb:626:in `block in run_around_example_hooks_for'",
 "/Users/apeque01/.rvm/gems/ruby-3.0.0/gems/rspec-core-3.13.0/lib/rspec/core/example.rb:352:in `call'",
 "/Users/apeque01/.rvm/gems/ruby-3.0.0/gems/rspec-expectations-3.13.1/lib/rspec/expectations/failure_aggregator.rb:25:in `block in aggregate'",
 "/Users/apeque01/.rvm/gems/ruby-3.0.0/gems/rspec-support-3.13.1/lib/rspec/support.rb:126:in `with_failure_notifier'",
 "/Users/apeque01/.rvm/gems/ruby-3.0.0/gems/rspec-expectations-3.13.1/lib/rspec/expectations/failure_aggregator.rb:23:in `aggregate'",
 "/Users/apeque01/.rvm/gems/ruby-3.0.0/gems/rspec-expectations-3.13.1/lib/rspec/matchers.rb:306:in `aggregate_failures'",
 "/Users/apeque01/.rvm/gems/ruby-3.0.0/gems/rspec-core-3.13.0/lib/rspec/core/configuration.rb:2281:in `block in define_built_in_hooks'",
 "/Users/apeque01/.rvm/gems/ruby-3.0.0/gems/rspec-core-3.13.0/lib/rspec/core/example.rb:457:in `instance_exec'",
 "/Users/apeque01/.rvm/gems/ruby-3.0.0/gems/rspec-core-3.13.0/lib/rspec/core/example.rb:457:in `instance_exec'",
 "/Users/apeque01/.rvm/gems/ruby-3.0.0/gems/rspec-core-3.13.0/lib/rspec/core/hooks.rb:390:in `execute_with'",
 "/Users/apeque01/.rvm/gems/ruby-3.0.0/gems/rspec-core-3.13.0/lib/rspec/core/hooks.rb:628:in `block (2 levels) in run_around_example_hooks_for'",
 "/Users/apeque01/.rvm/gems/ruby-3.0.0/gems/rspec-core-3.13.0/lib/rspec/core/example.rb:352:in `call'",
 "/Users/apeque01/.rvm/gems/ruby-3.0.0/gems/rspec-core-3.13.0/lib/rspec/core/hooks.rb:629:in `run_around_example_hooks_for'",
 "/Users/apeque01/.rvm/gems/ruby-3.0.0/gems/rspec-core-3.13.0/lib/rspec/core/hooks.rb:486:in `run'",
 "/Users/apeque01/.rvm/gems/ruby-3.0.0/gems/rspec-core-3.13.0/lib/rspec/core/example.rb:468:in `with_around_example_hooks'",
 "/Users/apeque01/.rvm/gems/ruby-3.0.0/gems/rspec-core-3.13.0/lib/rspec/core/example.rb:511:in `with_around_and_singleton_context_hooks'",
 "/Users/apeque01/.rvm/gems/ruby-3.0.0/gems/rspec-core-3.13.0/lib/rspec/core/example.rb:259:in `run'",
 "/Users/apeque01/.rvm/gems/ruby-3.0.0/gems/rspec-core-3.13.0/lib/rspec/core/example_group.rb:646:in `block in run_examples'",
 "/Users/apeque01/.rvm/gems/ruby-3.0.0/gems/rspec-core-3.13.0/lib/rspec/core/example_group.rb:642:in `map'",
 "/Users/apeque01/.rvm/gems/ruby-3.0.0/gems/rspec-core-3.13.0/lib/rspec/core/example_group.rb:642:in `run_examples'",
 "/Users/apeque01/.rvm/gems/ruby-3.0.0/gems/rspec-core-3.13.0/lib/rspec/core/example_group.rb:607:in `run'",
 "/Users/apeque01/.rvm/gems/ruby-3.0.0/gems/rspec-core-3.13.0/lib/rspec/core/example_group.rb:608:in `block in run'",
 "/Users/apeque01/.rvm/gems/ruby-3.0.0/gems/rspec-core-3.13.0/lib/rspec/core/example_group.rb:608:in `map'",
 "/Users/apeque01/.rvm/gems/ruby-3.0.0/gems/rspec-core-3.13.0/lib/rspec/core/example_group.rb:608:in `run'",
 "/Users/apeque01/.rvm/gems/ruby-3.0.0/gems/rspec-core-3.13.0/lib/rspec/core/runner.rb:121:in `block (3 levels) in run_specs'",
 "/Users/apeque01/.rvm/gems/ruby-3.0.0/gems/rspec-core-3.13.0/lib/rspec/core/runner.rb:121:in `map'",
 "/Users/apeque01/.rvm/gems/ruby-3.0.0/gems/rspec-core-3.13.0/lib/rspec/core/runner.rb:121:in `block (2 levels) in run_specs'",
 "/Users/apeque01/.rvm/gems/ruby-3.0.0/gems/rspec-core-3.13.0/lib/rspec/core/configuration.rb:2091:in `with_suite_hooks'",
 "/Users/apeque01/.rvm/gems/ruby-3.0.0/gems/rspec-core-3.13.0/lib/rspec/core/runner.rb:116:in `block in run_specs'",
 "/Users/apeque01/.rvm/gems/ruby-3.0.0/gems/rspec-core-3.13.0/lib/rspec/core/reporter.rb:74:in `report'",
 "/Users/apeque01/.rvm/gems/ruby-3.0.0/gems/rspec-core-3.13.0/lib/rspec/core/runner.rb:115:in `run_specs'",
 "/Users/apeque01/.rvm/gems/ruby-3.0.0/gems/rspec-core-3.13.0/lib/rspec/core/runner.rb:89:in `run'",
 "/Users/apeque01/.rvm/gems/ruby-3.0.0/gems/rspec-core-3.13.0/lib/rspec/core/runner.rb:71:in `run'",
 "/Users/apeque01/.rvm/gems/ruby-3.0.0/gems/rspec-core-3.13.0/lib/rspec/core/runner.rb:45:in `invoke'",
 "/Users/apeque01/.rvm/gems/ruby-3.0.0/gems/rspec-core-3.13.0/exe/rspec:4:in `<top (required)>'",
 "/Users/apeque01/.rvm/gems/ruby-3.0.0/bin/rspec:25:in `load'",
 "/Users/apeque01/.rvm/gems/ruby-3.0.0/bin/rspec:25:in `<main>'"]
 ```

Reply:
Ok, for our case that looks reasonable. @titusfortner Do you know if Sauce or remote server implementations might provide extra information in the backtrace that needs to be parsed?